File: 2024/q01.py
#!/usr/bin/env python


def part1(data):
    ans = data.count('B') + 3 * data.count('C')
    return ans


def part2(data):
    data = data.strip()
    def calc(s):
        return s.count('B') + 3 * s.count('C') + 5 * s.count('D')
    ans = 0
    for a,b in zip(data[::2], data[1::2]):
        ans += calc(a) + calc(b)
        if 'x' not in (a,b):
            ans += 2
    return ans


def part3(data):
    data = data.strip()
    def calc(s):
        return s.count('B') + 3 * s.count('C') + 5 * s.count('D')
    ans = 0
    for a,b,c in zip(data[::3], data[1::3], data[2::3]):
        ans += calc(a) + calc(b) + calc(c)
        match (a+b+c).count('x'):
            case 0:
                ans += 6
            case 1:
                ans += 2
    return ans


data = 'ABBAC'
assert part1(data) == 5

data = 'AxBCDDCAxD'
assert part2(data) == 28

data = 'xBxAAABCDxCC'
assert part3(data) == 30


data = open('q01_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q01_p2.txt').read()
ans = part2(data)
print('part2:', ans)

data = open('q01_p3.txt').read()
ans = part3(data)
print('part3:', ans)

File: event/2024/q01.jl
#!/usr/bin/env julia


function part1(data)
    sum(data) do c
        -1 + findfirst(c, "AB-C")
    end
end


function part2(data)
    vals = Dict(zip("xABCD", [0,0,1,3,5]))
    sum(Iterators.partition(data, 2)) do s
        x = 2('x' ∉ s)
        x + sum(c->vals[c], s)
    end
end


function part3(data)
    vals = Dict(zip("xABCD", [0,0,1,3,5]))
    sum(Iterators.partition(data, 3)) do s
        x = [6,2,0,0][begin+count('x', s)]
        x + sum(c->vals[c], s)
    end
end


@assert part1("ABBAC") == 5
@assert part2("AxBCDDCAxD") == 28
@assert part3("xBxAAABCDxCC") == 30


data = readchomp("q01_p1.txt")
println("part1: ", part1(data))
data = readchomp("q01_p2.txt")
println("part2: ", part2(data))
data = readchomp("q01_p3.txt")
println("part3: ", part3(data))
